### PR TITLE
Allow specifying host port to bind to

### LINKED ISSUE #314: Features to specify the host port in port mapping
I use testcontainers for running my integration tests. On my work computer, a lot of port ranges are blocked. That causes the `docker-modem` library to fail with the following error on around every third run of tests:

```
/home/abdullah/abb/WB-IdentityServer/node_modules/docker-modem/lib/modem.js:354
      callback(null, json);
      ^
Error: (HTTP code 500) server error - Ports are not available: listen tcp 0.0.0.0:58773: bind: An attempt was made to access a socket in a way forbidden by its access permissions.
 
    at /home/abdullah/abb/WB-IdentityServer/node_modules/docker-modem/lib/modem.js:324:17
    at getCause (/home/abdullah/abb/WB-IdentityServer/node_modules/docker-modem/lib/modem.js:354:7)
    at Modem.buildPayload (/home/abdullah/abb/WB-IdentityServer/node_modules/docker-modem/lib/modem.js:323:5)
    at IncomingMessage.<anonymous> (/home/abdullah/abb/WB-IdentityServer/node_modules/docker-modem/lib/modem.js:295:14)
    at IncomingMessage.emit (node:events:402:35)
    at IncomingMessage.emit (node:domain:475:12)
    at endReadableNT (node:internal/streams/readable:1343:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
```

A possible solution is to add to able to specify a host port as well for each port mapping e.g. in the `GenericContainer.withExposedPorts` method.

### PR DESCRIPTION
**Pull request by codetheweb**
Closes #314.

Open to renaming `PortWithOptionalBinding` if there's something else that makes more sense. :)

### CODE REVIEW

File: src/port.ts [hunk 1]
```diff
@@ -1,2 +1,12 @@
 export type Port = number;
 export type PortString = string;
+
+export type PortWithOptionalBinding =
+  | Port
```
Comment (cristianrgreco) [inline src/port.ts:5]: What's the trailing `|` do? Also below `number` -> `Port`

**Comment by cristianrgreco** [comment]
Thanks for raising @codetheweb. Could you please add an example to the README on how to use it, preferably with some disclaimer that it should be avoided where possible?

File: src/port.ts (hunk 1 above)
Comment (codetheweb) [reply src/port.ts:5]: Partly fixed in [8777232](https://github.com/testcontainers/testcontainers-node/pull/328/commits/8777232642285c9060c2a51f8a5604b0fba98b39). Prettier is adding the leading `|`.

**Comment by codetheweb** [comment]
Added an example to the readme. Thanks for the quick review. 😃

File: README.md [hunk 2]
```diff
@@ -433,6 +433,20 @@ const { output } = await container.exec(["curl", `http://host.testcontainers.int
 assert(output === "hello world");
 ```
 
+Specifying an exact host port to bind to (not recommended, Testcontainers will automatically find an available port):
+
+```javascript
+const { GenericContainer, Network } = require("testcontainers");
```
Comment (cristianrgreco) [inline README.md:439]: could you please remove the network part from this example?

File: src/docker/functions/container/create-container.ts [hunk 3]
```diff
@@ -79,10 +79,14 @@ const getExtraHosts = (extraHosts: ExtraHost[]): string[] => {
   return extraHosts.map((extraHost) => `${extraHost.host}:${extraHost.ipAddress}`);
 };
 
-const getPortBindings = (exposedPorts: Port[]): DockerodePortBindings => {
+const getPortBindings = (exposedPorts: PortWithOptionalBinding[]): DockerodePortBindings => {
   const dockerodePortBindings: DockerodePortBindings = {};
   for (const exposedPort of exposedPorts) {
-    dockerodePortBindings[exposedPort] = [{ HostPort: "0" }];
+    if (typeof exposedPort === "number") {
```
Comment (cristianrgreco) [inline src/docker/functions/container/create-container.ts:85]: in the same way that we have `getContainerPort` which does the type checking, perhaps we could also add `getHostPort`? that way we fully encapsulate the type differences

Comment (codetheweb) [reply src/docker/functions/container/create-container.ts:85]: I think that makes sense. Should it just return null if `typeof port === 'number'`?

File: README.md (hunk 2 above)
Comment (codetheweb) [reply README.md]: My bad, fixed in 6360c4b.

File: src/docker/functions/container/create-container.ts (hunk 3 above)
Comment (codetheweb) [reply src/docker/functions/container/create-container.ts]: Ah, I see. Added in a27f11b.

File: src/generic-container/generic-container.ts [hunk 4]
```diff
@@ -206,12 +206,12 @@ export class GenericContainer implements TestContainer {
     return this;
   }
 
-  public withExposedPorts(...ports: Port[]): this {
+  public withExposedPorts(...ports: PortWithOptionalBinding[]): this {
```
Comment (cristianrgreco) [inline src/generic-container/generic-container.ts:209]: the base interface `testcontainer.ts` also needs to be updated

**Comment by cristianrgreco** [comment]
last comment from me, thanks for addressing the feedback so quickly

File: src/generic-container/generic-container.ts (hunk 4 above)
Comment (codetheweb) [reply src/generic-container/generic-container.ts:209]: Fixed in 88d7ca8.

**Review by cristianrgreco — APPROVED** [review]

**Comment by cristianrgreco** [comment]
@codetheweb released in 8.4.0, thanks again

**Comment by ab316** [comment]
Thank you guys for fixing this :)

**Pull request merged**